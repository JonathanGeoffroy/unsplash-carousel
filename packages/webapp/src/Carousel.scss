@import "./vars";

.Carousel {
  max-width: 900px;
  position: relative;

  display: flex;
  flex-direction: column;

  button {
    border: none;
    background-color: transparent;
    font-family: inherit;
    padding: 0;
    cursor: pointer;

    @media screen and (-ms-high-contrast: active) {
      border: 2px solid currentcolor;
    }
  }

  > *:not(:last-child) {
    margin-bottom: 0.5em;
  }
}
