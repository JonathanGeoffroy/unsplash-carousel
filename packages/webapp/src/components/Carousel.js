import { createContext } from "react";
import useImages from "../hooks/useImages";
import Main from "./Main";
import Thumbnails from "./Thumbnails";

import "./Carousel.scss";

export const CarouselContext = createContext();

function Carousel() {
  const images = useImages();
  const { initialized } = images.state;

  return (
    <div className="Carousel">
      <CarouselContext.Provider value={images}>
        {initialized ? (
          <>
            <Main />
            <Thumbnails />
          </>
        ) : (
          <div className="skeleton">Loading ...</div>
        )}
      </CarouselContext.Provider>
    </div>
  );
}

export default Carousel;
