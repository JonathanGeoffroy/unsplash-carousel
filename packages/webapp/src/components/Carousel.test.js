import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import Carousel from "./Carousel";

test("Display loading until data is initialized", () => {
  render(<Carousel />);

  expect(screen.getByText(/Loading/g).textContent).toBe("Loading ...");

  return waitFor(() => screen.getByAltText("first description"));
});

test("Display Carousel when data is initialized", async () => {
  const { container } = render(<Carousel />);

  await waitFor(() => screen.getByAltText("first description"));

  expect(screen.getByAltText("first description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);

  expect(container).toMatchSnapshot();
});

test("Manage image selection button", async () => {
  const { container } = render(<Carousel />);

  await waitFor(() => screen.getByAltText("first description"));

  expect(screen.getByAltText("first description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);

  const nextButton = screen.getByLabelText(/Next image/g);
  fireEvent.click(nextButton);

  await waitFor(() => screen.getByAltText("second description"));
  expect(screen.queryByAltText("first description")).not.toBeInTheDocument();
  expect(screen.getByAltText("second description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);
  expect(container).toMatchSnapshot();

  fireEvent.click(nextButton);
  await waitFor(() => screen.getByAltText("first description"));
  expect(screen.queryByAltText("second description")).not.toBeInTheDocument();
  expect(screen.getByAltText("first description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);
  expect(container).toMatchSnapshot();

  const previousButton = screen.getByLabelText(/Previous image/g);
  fireEvent.click(previousButton);

  await waitFor(() => screen.getByAltText("second description"));
  expect(screen.queryByAltText("first description")).not.toBeInTheDocument();
  expect(screen.getByAltText("second description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);
  expect(container).toMatchSnapshot();

  fireEvent.click(previousButton);

  await waitFor(() => screen.getByAltText("first description"));
  expect(screen.queryByAltText("second description")).not.toBeInTheDocument();
  expect(screen.getByAltText("first description")).toBeInTheDocument();
  expect(screen.getAllByRole("img")).toHaveLength(3);
  expect(container).toMatchSnapshot();
});

test("Manage thumbnails selection", async () => {
  const { container } = render(<Carousel />);

  await waitFor(() => screen.getByAltText("first description"));
  expect(screen.getByAltText("first description")).toBeInTheDocument();

  const thumbnails = screen.getByTestId("thumbnails");
  const thumbnailButtons = within(thumbnails).getAllByRole("button");
  expect(thumbnailButtons).toHaveLength(2);

  fireEvent.click(thumbnailButtons[1]);
  await waitFor(() => screen.getByAltText("second description"));
  expect(screen.queryByAltText("first description")).not.toBeInTheDocument();
  expect(screen.getByAltText("second description")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
