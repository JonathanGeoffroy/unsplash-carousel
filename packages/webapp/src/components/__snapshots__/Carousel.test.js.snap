// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Carousel when data is initialized 1`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              first username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            first description
          </div>
          <div
            class="Image-keywords"
          >
            #first, #key, #word
          </div>
        </div>
        <img
          alt="first description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
      <button>
        <img
          alt=""
          class=""
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Manage image selection button 1`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              second username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            second description
          </div>
          <div
            class="Image-keywords"
          >
            #second, #key, #word
          </div>
        </div>
        <img
          alt="second description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class=""
        />
      </button>
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Manage image selection button 2`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              first username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            first description
          </div>
          <div
            class="Image-keywords"
          >
            #first, #key, #word
          </div>
        </div>
        <img
          alt="first description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
      <button>
        <img
          alt=""
          class=""
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Manage image selection button 3`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              second username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            second description
          </div>
          <div
            class="Image-keywords"
          >
            #second, #key, #word
          </div>
        </div>
        <img
          alt="second description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class=""
        />
      </button>
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Manage image selection button 4`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              first username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            first description
          </div>
          <div
            class="Image-keywords"
          >
            #first, #key, #word
          </div>
        </div>
        <img
          alt="first description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
      <button>
        <img
          alt=""
          class=""
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Manage thumbnails selection 1`] = `
<div>
  <div
    class="Carousel"
  >
    <div
      class="Main"
    >
      <button
        aria-label="Previous image"
        class="selector-button selector-button-previous"
      >
        ‹
      </button>
      <div
        class="Image"
      >
        <div
          class="Image-info"
        >
          <div
            class="Image-top"
          >
            <span
              class="Image-top-username"
            >
              second username
            </span>
            <span>
              01/01/2020
            </span>
          </div>
          <div
            class="Image-description"
          >
            second description
          </div>
          <div
            class="Image-keywords"
          >
            #second, #key, #word
          </div>
        </div>
        <img
          alt="second description"
          height="600"
          width="900"
        />
      </div>
      <button
        aria-label="Next image"
        class="selector-button selector-button-next"
      >
        ›
      </button>
      <div
        class="Action"
      >
        <button>
          ▶
        </button>
        <div
          class="timeline"
          style="width: 0%;"
        />
      </div>
    </div>
    <div
      class="thumbnails"
      data-testid="thumbnails"
    >
      <button>
        <img
          alt=""
          class=""
        />
      </button>
      <button>
        <img
          alt=""
          class="selected"
        />
      </button>
    </div>
  </div>
</div>
`;
